File: web/src/services/auth.service.ts
import { api } from './api';
import { AuthResponse, LoginRequest, RegisterRequest, User } from '@/types';

export const authService = {
  async login(credentials: LoginRequest): Promise<AuthResponse> {
    const response = await api.post<AuthResponse>('/auth/login', credentials);
    this.saveTokens(response);
    return response;
  },

  async register(data: RegisterRequest): Promise<AuthResponse> {
    const response = await api.post<AuthResponse>('/auth/register', data);
    this.saveTokens(response);
    return response;
  },

  async logout(): Promise<void> {
    try {
      await api.post('/auth/logout');
    } finally {
      this.clearTokens();
    }
  },

  async getProfile(): Promise<User> {
    const response = await api.get<{ user: User }>('/auth/profile');
    return response.user;
  },

  saveTokens(response: AuthResponse) {
    localStorage.setItem('accessToken', response.accessToken);
    localStorage.setItem('refreshToken', response.refreshToken);
    localStorage.setItem('userId', response.user.id);
  },

  clearTokens() {
    localStorage.removeItem('accessToken');
    localStorage.removeItem('refreshToken');
    localStorage.removeItem('userId');
  },

  isAuthenticated(): boolean {
    return !!localStorage.getItem('accessToken');
  },

  getAccessToken(): string | null {
    return localStorage.getItem('accessToken');
  },
};

File: ios/RecipeApp/Services/AuthService.swift
import Foundation

class AuthService {
    static let shared = AuthService()
    private let client = APIClient.shared

    private init() {}

    // MARK: - Authentication

    func register(email: String, password: String, username: String) async throws -> AuthResponse {
        let request = RegisterRequest(email: email, password: password, username: username)
        let response: AuthResponse = try await client.request(
            endpoint: "/auth/register",
            method: "POST",
            body: request,
            requiresAuth: false
        )

        // Save tokens
        KeychainService.shared.saveAccessToken(response.accessToken)
        KeychainService.shared.saveRefreshToken(response.refreshToken)
        KeychainService.shared.saveUserId(response.user.id.uuidString)

        return response
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        let request = LoginRequest(email: email, password: password)
        let response: AuthResponse = try await client.request(
            endpoint: "/auth/login",
            method: "POST",
            body: request,
            requiresAuth: false
        )

        // Save tokens
        KeychainService.shared.saveAccessToken(response.accessToken)
        KeychainService.shared.saveRefreshToken(response.refreshToken)
        KeychainService.shared.saveUserId(response.user.id.uuidString)

        return response
    }

    func logout() async throws {
        do {
            let _: EmptyResponse = try await client.request(
                endpoint: "/auth/logout",
                method: "POST"
            )
        } catch {
            // Continue with logout even if server request fails
            print("Logout request failed: \(error)")
        }

        // Clear local tokens
        KeychainService.shared.clearAll()
    }

    func getProfile() async throws -> User {
        struct ProfileResponse: Codable {
            let user: User
        }

        let response: ProfileResponse = try await client.request(
            endpoint: "/auth/profile",
            method: "GET"
        )

        return response.user
    }

    // MARK: - Token Management

    func isAuthenticated() -> Bool {
        return KeychainService.shared.getAccessToken() != nil
    }
}

struct EmptyResponse: Codable {}
